iree/compiler/Conversion/LinalgToLLVM/LinalgTileAndDistributePass.cpp

```diff
@@ -0,0 +1,236 @@
+
+  LogicalResult matchAndRewrite(Operation *op,
+                                PatternRewriter &rewriter) const override {
+    // Find the parent FuncOp before tiling. If tiling succeeds, the op will be
+    // erased.
+    FuncOp funcOp = op->getParentOfType<FuncOp>();
+    SmallVector<Value, 4> tensorResults;
+    if (!funcOp ||
+        failed(Base::matchAndRewriteBase(op, rewriter, tensorResults)) ||
+        !tensorResults.empty() ||
+        (funcOp.getAttr(getNumWorkgroupsFnAttrName()) &&
+         failed(utils::createNumWorkgroupsFromResultShape(
+             rewriter, cast<linalg::LinalgOp>(op), funcOp,
+             getNumWorkgroupsFnAttrName(),
+             cpuKernelDispatch.getTileSizes(op))))) {
+      return failure();
+    }
+    rewriter.eraseOp(op);
+    return success();
+  }
+  CPUKernelDispatch cpuKernelDispatch;
+};
+
+template <typename LinalgOpTy>
+struct TileAndFuseToCPUThreads
+    : public linalg::LinalgTileAndFusePattern<LinalgOpTy> {
+  using Base = linalg::LinalgTileAndFusePattern<LinalgOpTy>;
+  TileAndFuseToCPUThreads(MLIRContext *context,
+                          const linalg::LinalgDependenceGraph &dependenceGraph,
+                          const CPUKernelDispatch &cpuKernelDispatch,
+                          linalg::LinalgTilingOptions tilingOptions,
+                          linalg::LinalgMarker marker,
+                          PatternBenefit benefit = 1)
+      : Base(context, dependenceGraph, tilingOptions,
+             linalg::LinalgFusionOptions().setIndicesToFuse({2}), marker,
+             marker,
+             linalg::LinalgMarker(
+                 ArrayRef<Identifier>(),
+                 Identifier::get(getCPUDeleteMarker(), context)),
+             benefit),
+        dependenceGraph(dependenceGraph),
+        cpuKernelDispatch(cpuKernelDispatch) {}
+
+  virtual LogicalResult matchAndRewrite(Operation *op,
+                                        PatternRewriter &rewriter) const {
+    FuncOp funcOp = op->getParentOfType<FuncOp>();
+    linalg::LinalgOp linalgOp = cast<linalg::LinalgOp>(op);
+    if (!funcOp || !dependenceGraph.hasDependentOperations(linalgOp) ||
+        failed(Base::matchAndRewrite(op, rewriter)) ||
+        failed(utils::createNumWorkgroupsFromResultShape(
+            rewriter, cast<linalg::LinalgOp>(op), funcOp,
+            getNumWorkgroupsFnAttrName().str(),
+            cpuKernelDispatch.getTileSizes(op)))) {
+      return failure();
+    }
+    return success();
+  }
+
+  const linalg::LinalgDependenceGraph &dependenceGraph;
+  CPUKernelDispatch cpuKernelDispatch;
+};
+
+// We use |tiles| = |threads| no need to have iree.workgroup_count op.
+class RemoveNumThreads : public OpRewritePattern<IREE::WorkgroupCountOp> {
+ public:
+  using OpRewritePattern<IREE::WorkgroupCountOp>::OpRewritePattern;
+  LogicalResult matchAndRewrite(IREE::WorkgroupCountOp op,
+                                PatternRewriter &rewriter) const override {
+    rewriter.eraseOp(op);
```

Comment:
You are erasing the op and returning failure? If the op is just dead, then adding an NoSideEffect attribute in ODS will make CSE remove it. I dont think you need a specific pattern

Reply:
Done! thanks thats better =)